### PR TITLE
Adding telemetry report functions for DNC.

### PR DESCRIPTION
**Pull request by jaer-tsun**
**What this PR does / why we need it**:
Adding telemetry report functionality for DNC.

### CODE REVIEW

File: telemetry/telemetry.go [hunk 1]
```diff
@@ -256,6 +293,38 @@ func (reportMgr *NPMReportManager) SendReport() error {
...
 }
 
+// SendReport will send DNC telemetry report to HostNetAgent.
+func (reportMgr *DNCReportManager) SendReport() error {
+	var body bytes.Buffer
+
+	httpc := &http.Client{}
+	json.NewEncoder(&body).Encode(reportMgr.Report)
+
+	log.Printf("Going to send Telemetry report to hostnetagent %v", reportMgr.ReportManager.HostNetAgentURL)
```
Comment (tamilmani1989) [inline telemetry/telemetry.go]: move this to common,.i see lot of things in common in other send report functions

File: telemetry/telemetry.go [hunk 2]
```diff
@@ -124,10 +147,19 @@ type NPMReportManager struct {
...
+}
+
 const (
-	// TelemetryFile Path.
+	// CNITelemetryFile Path.
 	CNITelemetryFile = platform.CNIRuntimePath + "AzureCNITelemetry.json"
+	// NPMTelemetryFile Path.
 	NPMTelemetryFile = platform.NPMRuntimePath + "AzureNPMTelemetry.json"
+	// DNCTelemetryFile Path.
+	DNCTelemetryFile = platform.DNCRuntimePath + "AzureDNCTelemetry.json"
```
Comment (tamilmani1989) [inline telemetry/telemetry.go]: make sure you are using this..

**Comment by tamilmani1989** [comment]
There are some design changes required. As I see few functions are redundant.

File: telemetry/telemetry.go [hunk 3]
```diff
@@ -417,22 +431,99 @@ func (report *CNIReport) GetInterfaceDetails(queryUrl string) {
...
 	} else {
-		report.OrchestratorDetails.ErrorMessage = "Length of array is less than 2"
+		err = fmt.Errorf("[Telemetry] Response body is empty")
 	}
+
+	return err
+}
+
+// GetSystemDetails - retrieve system details like cpu usage, mem usage, and number or running processes
+func (reportMgr *ReportManager) GetSystemDetails() error {
```
Comment (saiyan86) [inline telemetry/telemetry.go]: Check with CLA. Remove it for now.

File: telemetry/telemetry.go [hunk 4]
```diff
@@ -93,7 +139,10 @@ type ClusterState struct {
 
 // NPMReport structure.
 type NPMReport struct {
-	StartFlag         bool
```
Comment (saiyan86) [inline telemetry/telemetry.go]: add metadata

Comment (saiyan86) [inline telemetry/telemetry.go]: add metadata

File: telemetry/telemetry.go [hunk 5]
```diff
@@ -67,9 +76,50 @@ type OrchestratorInfo struct {
...
+	OsType               string `json:"osType"`
+	PlacementGroupID     string `json:"placementGroupId"`
+	PlatformFaultDomain  string `json:"platformFaultDomain"`
+	PlatformUpdateDomain string `json:"platformUpdateDomain"`
+	Publisher            string `json:"publisher"`
+	ResourceGroupName    string `json:"resourceGroupName"`
+	Sku                  string `json:"sku"`
+	SubscriptionID       string `json:"subscriptionId"`
+	Tags                 string `json:"tags"`
+	Version              string `json:"version"`
```
[the remark this answers is not in the document]
Comment (saiyan86) [reply telemetry/telemetry.go]: Add KernelVersion

File: telemetry/telemetry.go [hunk 6]
```diff
@@ -67,9 +76,50 @@ type OrchestratorInfo struct {
...
+	NPMReportType = "NPM"
+	// DNC report type
+	DNCReportType = "DNC"
+)
+
+// Metadata retrieved from wireserver
+type Metadata struct {
+	Location             string `json:"location"`
+	VMName               string `json:"name"`
+	Offer                string `json:"offer"`
```
Comment (saiyan86) [inline telemetry/telemetry.go]: add a comment //Offer is the OS distribution.

File: telemetry/telemetry.go [hunk 7]
```diff
@@ -417,22 +431,99 @@ func (report *CNIReport) GetInterfaceDetails(queryUrl string) {
...
+				v.FieldByName("PlatformFaultDomain").SetString(report.Metadata.PlatformFaultDomain)
+				v.FieldByName("PlatformUpdateDomain").SetString(report.Metadata.PlatformUpdateDomain)
+				v.FieldByName("Publisher").SetString(report.Metadata.Publisher)
+				v.FieldByName("ResourceGroupName").SetString(report.Metadata.ResourceGroupName)
+				v.FieldByName("Sku").SetString(report.Metadata.Sku)
+				v.FieldByName("SubscriptionID").SetString(report.Metadata.SubscriptionID)
+				v.FieldByName("Tags").SetString(report.Metadata.Tags)
+				v.FieldByName("Version").SetString(report.Metadata.Version)
+				v.FieldByName("VMID").SetString(report.Metadata.VMID)
+				v.FieldByName("VMSize").SetString(report.Metadata.VMSize)
```
Comment (saiyan86) [inline telemetry/telemetry.go]: set kernel version here

**Pull request merged**